### ISSUE TITLE
Raft.db Inflates - Performance Degradation 

### ISSUE THREAD

**Issue by orarnon** [hashicorp/consul#5181]
#### Overview of the Issue

Yesterday, we've reached a raft.db size of 1.7GB and were effectively down. Only the deletion of raft library helped to recover Consul.
The file started at 100MB and inflated since then to 340MB + 107MB in snapshots.
We're looking for ways to investigate why we've reached this size and why the DB is inflated. 

### Consul info for both Client and Server

<details>
  <summary>Server info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 3
	checks = 3
	services = 4
build:
	prerelease =
	revision = 0bddfa23
	version = 1.4.0
consul:
	acl = disabled
	bootstrap = false
	known_datacenters = 0
	leader = false
	leader_addr = 10.200.4.70:8300
	server = true
raft:
	applied_index = 481127591
	commit_index = 481127591
	fsm_pending = 0
	last_contact = 13.004652ms
	last_log_index = 481127591
	last_log_term = 82449
	last_snapshot_index = 481127449
	last_snapshot_term = 82449
	latest_configuration = [{Suffrage:Voter ID:673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 Address:10.200.1.154:8300} {Suffrage:Voter ID:77d2bb93-7859-2846-e090-2050e72768c8 Address:10.200.4.70:8300} {Suffrage:Voter ID:6568e22c-2660-bdb8-8d47-5ffc65c1601d Address:10.200.10.157:8300}]
	latest_configuration_index = 480777639
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Follower
	term = 82449
runtime:
	arch = amd64
	cpu_count = 8
	goroutines = 59679
	max_procs = 4
	os = linux
	version = go1.11.1
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 15377
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 416
	member_time = 2384490
	members = 2072
	query_queue = 0
	query_time = 245
```

</details>

### Operating system and Environment details

root@consul3:/var/lib/consul/raft# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.5 LTS
Release:	14.04
Codename:	trusty

Running in AWS on a c5d.2xlarge instance type.

### Log Fragments

<details>
  <summary>Log</summary>

```
2019/01/02 12:34:35 [INFO] raft: Starting snapshot up to 481141933
2019/01/02 12:34:35 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481141933-1546432475010.tmp
2019/01/02 12:34:35 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481108754-1546429312665
2019/01/02 12:34:35 [INFO] raft: Compacting logs from 481115235 to 481131717
2019/01/02 12:34:35 [INFO] raft: Snapshot to 481141933 complete
2019/01/02 13:00:40 [INFO] raft: Starting snapshot up to 481158630
2019/01/02 13:00:40 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481158630-1546434040388.tmp
2019/01/02 13:00:40 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481125472-1546430863532
2019/01/02 13:00:40 [INFO] raft: Compacting logs from 481131718 to 481148397
2019/01/02 13:00:40 [INFO] raft: Snapshot to 481158630 complete
2019/01/02 13:12:02 [INFO] raft: Starting snapshot up to 481175689
2019/01/02 13:12:02 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481175689-1546434722241.tmp
2019/01/02 13:12:02 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481141933-1546432475010
2019/01/02 13:12:02 [INFO] raft: Compacting logs from 481148398 to 481165453
2019/01/02 13:12:02 [INFO] raft: Snapshot to 481175689 complete
2019/01/02 13:15:37 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.751080271s
2019/01/02 13:16:03 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.780538391s
2019/01/02 13:28:14 [INFO] raft: Starting snapshot up to 481192372
2019/01/02 13:28:14 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481192372-1546435694716.tmp
2019/01/02 13:28:16 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481158630-1546434040388
2019/01/02 13:28:16 [INFO] raft: Compacting logs from 481165454 to 481182163
2019/01/02 13:28:16 [INFO] raft: Snapshot to 481192372 complete
2019/01/02 13:30:37 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.507277753s
2019/01/02 13:55:16 [INFO] raft: Starting snapshot up to 481208907
2019/01/02 13:55:16 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481208907-1546437316333.tmp
2019/01/02 13:55:17 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481175689-1546434722241
2019/01/02 13:55:17 [INFO] raft: Compacting logs from 481182164 to 481198682
2019/01/02 13:55:17 [INFO] raft: Snapshot to 481208907 complete
2019/01/02 14:13:01 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.522051286s
2019/01/02 14:18:35 [INFO] raft: Starting snapshot up to 481225358
2019/01/02 14:18:35 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481225358-1546438715249.tmp
2019/01/02 14:18:35 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481192372-1546435694716
2019/01/02 14:18:35 [INFO] raft: Compacting logs from 481198683 to 481215120
2019/01/02 14:18:35 [INFO] raft: Snapshot to 481225358 complete
2019/01/02 14:22:58 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.729410446s
2019/01/02 12:34:35 [INFO] raft: Starting snapshot up to 481141933
2019/01/02 12:34:35 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481141933-1546432475010.tmp
2019/01/02 12:34:35 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481108754-1546429312665
2019/01/02 12:34:35 [INFO] raft: Compacting logs from 481115235 to 481131717
2019/01/02 12:34:35 [INFO] raft: Snapshot to 481141933 complete
2019/01/02 13:00:40 [INFO] raft: Starting snapshot up to 481158630
2019/01/02 13:00:40 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481158630-1546434040388.tmp
2019/01/02 13:00:40 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481125472-1546430863532
2019/01/02 13:00:40 [INFO] raft: Compacting logs from 481131718 to 481148397
2019/01/02 13:00:40 [INFO] raft: Snapshot to 481158630 complete
2019/01/02 13:12:02 [INFO] raft: Starting snapshot up to 481175689
2019/01/02 13:12:02 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481175689-1546434722241.tmp
2019/01/02 13:12:02 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481141933-1546432475010
2019/01/02 13:12:02 [INFO] raft: Compacting logs from 481148398 to 481165453
2019/01/02 13:12:02 [INFO] raft: Snapshot to 481175689 complete
2019/01/02 13:15:37 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.751080271s
2019/01/02 13:16:03 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.780538391s
2019/01/02 13:28:14 [INFO] raft: Starting snapshot up to 481192372
2019/01/02 13:28:14 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481192372-1546435694716.tmp
2019/01/02 13:28:16 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481158630-1546434040388
2019/01/02 13:28:16 [INFO] raft: Compacting logs from 481165454 to 481182163
2019/01/02 13:28:16 [INFO] raft: Snapshot to 481192372 complete
2019/01/02 13:30:37 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.507277753s
2019/01/02 13:55:16 [INFO] raft: Starting snapshot up to 481208907
2019/01/02 13:55:16 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481208907-1546437316333.tmp
2019/01/02 13:55:17 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481175689-1546434722241
2019/01/02 13:55:17 [INFO] raft: Compacting logs from 481182164 to 481198682
2019/01/02 13:55:17 [INFO] raft: Snapshot to 481208907 complete
2019/01/02 14:13:01 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.522051286s
2019/01/02 14:18:35 [INFO] raft: Starting snapshot up to 481225358
2019/01/02 14:18:35 [INFO] snapshot: Creating new snapshot at /var/lib/consul/raft/snapshots/82449-481225358-1546438715249.tmp
2019/01/02 14:18:35 [INFO] snapshot: reaping snapshot /var/lib/consul/raft/snapshots/82449-481192372-1546435694716
2019/01/02 14:18:35 [INFO] raft: Compacting logs from 481198683 to 481215120
2019/01/02 14:18:35 [INFO] raft: Snapshot to 481225358 complete
2019/01/02 14:22:58 [WARN] raft: Failed to contact 673e0ff3-815a-4c56-0b5b-1ccf17d7a7d1 in 1.729410446s
```

</details>

**Comment by leprechau** [comment]
We had a very similar issue a bit over a year ago. The raft log grew kept growing and compaction kept taking longer and longer to the point where we were having 2-3 second outages every time compaction triggered.  The solution was to trim all the health checks, eliminate all dynamic checks and then remove, clean data folder (rm -rf style), and re-join each node one at a time.

**Comment by orarnon** [comment]
@leprechau can you elaborate how to examine raft.db or how you have identified this issue?
Do you mean:
1. stop consul.
2. rm -rf raft directory
3. start consul
?

**Comment by pearkes** [comment]
Sorry you're seeing issues here. I'm going to close this in favor of https://github.com/hashicorp/consul/issues/866 as I believe it is more or less a duplicate of that issue.

I would point to two configuration parameters that are low-level but can be tweaked to fit the use-case you have:

- https://www.consul.io/docs/agent/options.html#_raft_snapshot_threshold
- https://www.consul.io/docs/agent/options.html#_raft_snapshot_interval

This gives more control over the snapshotting behavior and explains the trade-off you'd be making. I hope that helps.

**Comment by leprechau** [comment]
@orarnon The `raft.db` is a boltdb file but you can also do some rudimentary analysis just using `strings`.  When you run that against the `raft.db` you will see the output of all your health checks as they are committed to the log.  When you have very large health check payloads that change frequently those changes are committed to the log.  That can easily grow the size of the raft database very quickly.

In our case we had a health check that was returning a very large payload with dynamic information (timestamp) in the return.  That meant that every check inteval (5s in our case) was causing a massive commit to the raft log.  Once we streamlined all of our health checks and removed all dynamic content from the health check return we did basically as you guessed ...

**Comment by orarnon** [comment]
@leprechau can you share how you ran strings and analyzed it?

**Comment by banks** [comment]
@orarnon not a good solution necessarily but there are tools around like https://github.com/br0xen/boltbrowser that let you open a boltdb file and inspect it's contents. Whether or not it makes it easy for you to tell what is taking up the space is another matter since the logical contents you see through a browser and the physical file size on disk are not necessarily related.

In general boltDB's design is a single grow-only file. That means that if you add 1GB of log entries (e.g. by benchmarking lots of writes through Consul) and then take a snapshot which truncates/deletes them again, you will only see a small number of recent log entries in the file if you use a browser or write a program, but the actual file on disk will never shrink smaller than the 1GB size it had to grow to.

In general outside of extreme benchmarks this is tolerable, the only fix for it is to replace the raft log store with an actual log and not use boltDB at all. We've planned to do that for a very long time, but it's a huge task and this is about the only material benefit which doesn't affect most production use cases outside of pathological benchmarking conditions.

We still want to fix it, just difficult to prioritize that low-level work currently!

**Comment by banks** [comment]
The metrics that are most useful here would be FSM transaction rate and catalog RPC rates - check outputs that change often cause writes to the _catalog_ not the KV store.

Note also that there was a bug that was introduced in 1.2.3 which is fixed in master and will be in 1.4.1 that causes nodes to sync their state every AE interval even if it's not changed, that may or may not impact you. In general the issue with checks with dynamic output is that it causes the check output to be written back to the server every AE interval (2-15 mins depending on cluster size). Which is usually not too bad due to that slow pace, but if you have tens of thousands of those checks then you could end up writing quite a bit through raft for sure which might effect raft commit times.

I really doubt the Raft Db size has anything to do with slow raft commits though. The growth you showed is expected given grow-only nature of boltDB it seems both a reasonable size and stable so I don't think it's part of your problem here.

Based on that, I'd suggest you start a new thread in the mailing list to get help finding cause for performance issues. If you feel you have enough info to open a new issue for it then feel free but It woudl help to have a full description of your workload (how many clients, what kinds of requests, how often do services churn etc) as well as server resource utilisation (disk IO, RAM, CPU including io wait) etc. and all of the raft metrics and any others you see that you think are useful (like RPC rates for catalog/kv/etc.

Hope this helps!
